### PR TITLE
Remove SPLUNK_METRICS_ENDPOINT from Instrumentation Libraries

### PR DESCRIPTION
**Pull request by pellared**
## Why

Fixes https://github.com/signalfx/gdi-specification/issues/194

The `SPLUNK_METRICS_ENDPOINT` is either not a stable setting or it is not supported in any of the existing instrumentation libraries.

1. In https://github.com/signalfx/splunk-otel-java it is marked as "experimental"
2. In https://github.com/signalfx/splunk-otel-js it looks like a undocumented functionality (so even less than "experimental")
3. https://github.com/signalfx/splunk-otel-go logs that it is not supported if it is set

Initially I wanted to add a "note" that this env var is not required, but a lean towards removing it from the spec. We should use/follow OTel metrics.

Also, during GDI spec review, none of the @signalfx/gdi-specification-maintainers was considering this env var as MUST HAVE.

At last, we do not want to implement it in https://github.com/signalfx/splunk-otel-dotnet

## What 

- In Instrumentation Libraries move `SPLUNK_METRICS_ENDPOINT` from "stable" to "experimental" (and optional) and also introduce other env vars defined here: https://github.com/signalfx/splunk-otel-java/blob/main/docs/advanced-config.md
- Fix issues reported by markdownlint

### CODE REVIEW

**Review by pjanotti — APPROVED** [review]
Per @breedx-splk comment, it makes sense to list it somewhere else as experimental, but, not on the "stable" table. LGTM.

**Comment by pellared** [comment]
> There are 14 code uses in the signalfx org: https://github.com/search?q=org%3Asignalfx+SPLUNK_METRICS_ENDPOINT&type=code

I checked all of them. Take notice that some are related to "serverless"

> Java calls many of its configuration items experimental to set a clear expectation that the configuration names could change.
Java uses even env vars which are not 

Also see: https://github.com/signalfx/gdi-specification/blob/c9216b2d05faa82da4f5a62d801476994f56ff56/specification/configuration.md#L12-L19

I am not sure if the fact that Java has it is a reason to have it in GDI spec. Especially that Java has env vars that are NOT mentioned in the GDI spec (therefore it violates the spec).

Moreover, I think that right now I think that we SHOULD NOT implement `SPLUNK_METRICS_ENDPOINT` which is not inline with the current recommendation that says that experimental (e.g. profiling) SHOULD be supported.

This is the reason why I proposed to remove this from the spec.

Maybe the consensus would be to put it as "Deprecated". I could pull also other Java metric-related env vars so that it does not violate the GDI spec 😉 

@breedx-splk What do you propose? I do not have any strong opinion.

**Comment by breedx-splk** [comment]
> Especially that Java has env vars that are NOT mentioned in the GDI spec (therefore it violates the spec).

This is blatantly false. The spec is not intended to be comprehensive, and languages/distributions/instrumentations are allowed to have supplemental functionality and configuration.

> @breedx-splk What do you propose? I do not have any strong opinion. Do you think a "Deprecated" table would be the best way going forward?

I hadn't really intended to propose anything. We can choose to remove it, but Java will still need a way for users to specify where to send metrics.

**Comment by breedx-splk** [comment]
> The spec says something else 🤷

Wow. Do you think that is supposed to also include upstream configuration items? Cuz if so, I have about 99 issues to create....better get busy!

**Comment by pellared** [comment]
@breedx-splk PTAL: https://github.com/signalfx/gdi-specification/pull/196/commits/341a14e819b0fdb5d235efa770f4c0accda104fd

Upstream configurations are not needed. This is only about Splunk-sepcific configuration 😉

File: specification/configuration.md
Comment (pjanotti) [inline specification/configuration.md:240]:
```suggestion
(instrumentation libraries MAY support them):
```

File: specification/configuration.md
Comment (pjanotti) [inline specification/configuration.md:324]:
```suggestion
  - traces (all other cases):`https://ingest.${SPLUNK_REALM}.signalfx.com/v2/trace`
```

File: specification/configuration.md
Comment (pjanotti) [reply specification/configuration.md:240]: @breedx-splk are these "workaround" settings or something that eventually should reach the GDI spec?

**Review by pjanotti — APPROVED** [review]
The broad agreement is about removing the mentioned setting from stable. There is a big issue with the scope of GDI spec. Let's handle that as a separate issue.

**Comment by pjanotti** [comment]
@pellared @breedx-splk my first take is contrary to what is suggested at https://github.com/signalfx/gdi-specification/blob/c9216b2d05faa82da4f5a62d801476994f56ff56/specification/configuration.md?plain=1#L15-L19

The GDI spec should cover the minimal env vars required on all stable instrumentations. It seems unattainable to do otherwise. Anyway, it would be better to handle it separately from this review.

File: specification/configuration.md
Comment (pjanotti) [reply specification/configuration.md:240]: I'm having second thoughts if we want to document such config options here. Unless this is something that we think that eventually, all instrumentation should have, it will be easier to just remove them.

**Comment by seemk** [comment]
I agree that we should prefer OTel specific env vars, so for repositories that have moved to OTel metrics, this env var is not really necessary any longer.

In the JS distro the preference is `programmatically -> SPLUNK_METRICS_ENDPOINT -> (handled by OTel) OTEL_EXPORTER_OTLP_METRICS_ENDPOINT -> OTEL_EXPORTER_OTLP_ENDPOINT`.

`SPLUNK_METRICS_ENDPOINT` is a remnant of the SignalFx metrics SDK, not really necessary any longer so it could be deprecated as long as the SDK supports OTel fully?

The preferred endpoint setup should anyway (as long as OTel is supported for all signals) be:

* For direct to cloud:
```
SPLUNK_REALM=us0
SPLUNK_ACCESS_TOKEN=...
launch
```
* To collector:
```
OTEL_EXPORTER_OTLP_ENDPOINT=collector:4317
launch
```

**Comment by pellared** [comment]
I see 3 options (from the most to least aggressive):

1. remove this configuration from the spec (it was not said to be "stable" in any distro) - I think it was a bug that this env var ended up in the "stable" section - _this was the initial version of the PR_
2. deprecate the env var
3. move the env var to "optional experimental" configuration - _current PR version, as (based on existing comments) I think it is the easiest way to go forward_

**Comment by seemk** [comment]
I'm up for removing it from the spec and leave it to distro authors not to break user code based on whatever versioning rules for a given platform :stuck_out_tongue:

**Comment by pellared** [comment]
> > Especially that Java has env vars that are NOT mentioned in the GDI spec (therefore it violates the spec).
> 
> This is blatantly false. The spec is not intended to be comprehensive, and languages/distributions/instrumentations are allowed to have supplemental functionality and configuration.
> 
> > @breedx-splk What do you propose? I do not have any strong opinion. Do you think a "Deprecated" table would be the best way going forward?
> 
> I hadn't really intended to propose anything. We can choose to remove it, but Java will still need a way for users to specify where to send metrics.

@breedx-splk  I created https://github.com/signalfx/gdi-specification/issues/198

**Comment by mateuszrzeszutek** [comment]
> My preference is **1 (removal)**. However, I we also start with **3 (optional experimental)** if anyone would have objections.

I think we should remove these config options - even if some instrumentation libraries (java) still use them. `SPLUNK_METRICS_*` configs and the entire Java metrics implementation are just tech debt, I believe we should not be codifying this crap in any spec. Long term, in the (hopefully close) future we're going to be using OTel metrics everywhere.

**Comment by mateuszrzeszutek** [comment]
> The GDI spec should cover the minimal env vars required on all stable instrumentations. It seems unattainable to do otherwise.

Even OTel spec does that -- language SDKs/instrumentations have a ton of experimental configs that are not specced. Some of is is obviously going to change in the future (once we get some kind of general "instrumentation configuration" spec), but I think that we'll always have language/implementation specific settings.

**Review by mateuszrzeszutek — APPROVED** [review]

**Comment by pellared** [comment]
Based on feedback I decided to revert the PR to the original state: **removal**).

By @breedx-splk 

> We can choose to remove it, but Java will still need a way for users to specify where to send metrics.

By @seemk 

> I'm up for removing it from the spec and leave it to distro authors not to break user code based on whatever versioning rules for a given platform

By @mateuszrzeszutek 

> I think we should remove these config options - even if some instrumentation libraries (java) still use them. 

I also created https://github.com/signalfx/gdi-specification/issues/198 to address the issue of introducing language/implementation specific settings.

File: specification/configuration.md
Comment (pellared) [reply specification/configuration.md]: Per @mateuszrzeszutek comment: we do not want to have these settings. I reverted the PR to the original state.

**Review by seemk — APPROVED** [review]

**Review by pjanotti — APPROVED** [review]

**Comment by pellared** [comment]
@breedx-splk PTAL

**Review by breedx-splk — APPROVED** [review]
As it reads now (one line removal of the MUST for `SPLUNK_METRICS_ENDPOINT`) I think I'm good with this. The way I read this is that implementations will no longer be required to implement `SPLUNK_METRICS_ENDPOINT`, but to me that doesn't preclude implementations from having it (and yeah #198 helps with this).

Sorry if that was the original intention and I missed the point. I just wanted to make sure that removal of this didn't instantly make java "in violation". 

We've got other approvals as well, so let's do this and continue forward.

**Comment by theletterf** [comment]
@breedx-splk @seemk LMK if this will have an impact on our current config / get started docs. Currently, we're using `SPLUNK_METRICS_ENDPOINT` for Java and NodeJS.

**Pull request merged**
